File: firmware/components/sonification_math/sonification_math.cpp
#include "sonification_math.h"
#include "calc_engine.h"

#include <cmath>

double sonify_map_y_to_freq(double y, double y_min, double y_max) {
    double span = y_max - y_min;
    if (span == 0.0) span = 1.0;
    double yNorm = (y - y_min) / span;
    if (yNorm < 0.0) yNorm = 0.0;
    if (yNorm > 1.0) yNorm = 1.0;
    return 200.0 * std::pow(5.0, yNorm);
}

double sonify_map_x_to_pan(double x, double x_min, double x_max) {
    double span = x_max - x_min;
    if (span == 0.0) span = 1.0;
    double xNorm = (x - x_min) / span;
    if (xNorm < 0.0) xNorm = 0.0;
    if (xNorm > 1.0) xNorm = 1.0;
    return -1.0 + 2.0 * xNorm;
}

crit_point_t sonify_check_critical(double x, const expr_handle_t *expr, angle_mode_t mode) {
    crit_point_t r{CRIT_NONE, 0.0};
    if (!expr) return r;

    const double y = calc_evaluate_at(expr, x, mode);
    if (!std::isfinite(y)) return r;

    const double eps = 0.005;
    const double yPrev = calc_evaluate_at(expr, x - eps, mode);
    const double yNext = calc_evaluate_at(expr, x + eps, mode);
    if (!std::isfinite(yPrev) || !std::isfinite(yNext)) return r;

    if (y > yPrev && y > yNext) {
        r.type = CRIT_MAXIMUM;
        r.y = y;
        return r;
    }
    if (y < yPrev && y < yNext) {
        r.type = CRIT_MINIMUM;
        r.y = y;
        return r;
    }
    if (std::fabs(y) < 1e-5) {
        r.type = CRIT_ROOT;
        r.y = 0.0;
        return r;
    }
    return r;
}

File: SonificationMath.js
/**
 * SonificationMath.js
 * Pitch/pan mapping and critical-point detection. DOM-free.
 * Depends on CalcEngine.evaluateAt for angle-mode-aware evaluation.
 */

function resolveCalcEngine() {
  if (typeof CalcEngine !== 'undefined') return CalcEngine;
  if (typeof globalThis !== 'undefined' && globalThis.CalcEngine) return globalThis.CalcEngine;
  return require('./CalcEngine');
}

class SonificationMath {
  static mapYToFrequency(y, yMin, yMax) {
    // Exponential mapping from [yMin, yMax] to [200, 1000] Hz
    const yNorm = Math.min(Math.max((y - yMin) / (yMax - yMin), 0.0), 1.0);
    // f(y) = 200 * (1000 / 200) ^ yNorm
    return 200.0 * Math.pow(5.0, yNorm);
  }

  static mapXToPan(x, xMin, xMax) {
    // Linear mapping from [xMin, xMax] to [-1.0, 1.0]
    const xNorm = Math.min(Math.max((x - xMin) / (xMax - xMin), 0.0), 1.0);
    return -1.0 + 2.0 * xNorm;
  }

  /**
   * Detect local min/max or near-root at x using angle-mode-aware evaluation.
   * @param {number} x
   * @param {Object} compiled
   * @param {string} angleMode - 'rad' or 'deg'
   * @returns {{type: string, y: number}|null}
   */
  static checkCriticalPoint(x, compiled, angleMode = 'rad') {
    if (!compiled) return null;

    const CE = resolveCalcEngine();

    try {
      const y = CE.evaluateAt(compiled, x, angleMode);
      if (isNaN(y) || !isFinite(y)) return null;

      const eps = 0.005; // Small delta for derivative check
      const yPrev = CE.evaluateAt(compiled, x - eps, angleMode);
      const yNext = CE.evaluateAt(compiled, x + eps, angleMode);

      if (isNaN(yPrev) || isNaN(yNext) || !isFinite(yPrev) || !isFinite(yNext)) return null;

      // Local maximum
      if (y > yPrev && y > yNext) {
        return { type: 'maximum', y };
      }

      // Local minimum
      if (y < yPrev && y < yNext) {
        return { type: 'minimum', y };
      }

      // Zero crossing (root)
      if (Math.abs(y) < 1e-5) {
        return { type: 'root', y: 0 };
      }
    } catch {
      // Ignore evaluation errors
    }
    return null;
  }
}

if (typeof module !== 'undefined' && module.exports) {
  module.exports = SonificationMath;
} else {
  window.SonificationMath = SonificationMath;
}
